topics/rate-limit.md

```diff
@@ -0,0 +1,146 @@
+
+
+## Configure %plugin_name% {id="configure"}
+
+### Overview {id="overview"}
+
+Ktor uses the _token bucket_ algorithm for rate limiting, which works as follows:
+1. In the beginning, we have a bucket defined by its capacity - the number of tokens.
+2. Each incoming request tries to consume one token from the bucket:
+    - If there is enough capacity, the server handles a request and sends a response with the following headers:
+        - `X-RateLimit-Limit`: a specified bucket capacity.
+        - `X-RateLimit-Remaining`: the number of tokens remaining in a bucket.
+        - `X-RateLimit-Reset`: a UTC timestamp (in seconds) that specifies the time of refilling a bucket.
+    - If there is insufficient capacity, the server rejects a request using a `429 Too Many Requests` response and adds the `Retry-After` header, indicating how long the client should wait before making a follow-up request.
+3. After a specified period of time, a bucket capacity is refilled.
+
+
+### Register a rate limiter {id="register"}
+Ktor allows you to apply rate limiting globally to a whole application or to specific routes:
+- To apply rate limiting to a whole application, call the `global` method and pass a configured rate limiter.
+   ```kotlin
+   install(RateLimit) {
+       global {
+           rateLimiter(limit = 5, refillPeriod = 60.seconds)
+       }
+   }
+   ```
+
+- The `register` method registers a rate limiter that can be applied to specific routes.
+   ```kotlin
+   ```
+   {src="snippets/rate-limit/src/main/kotlin/com/example/Application.kt" include-lines="14-17,33"}
+
+Code samples above demonstrate minimal configurations for the `%plugin_name%` plugin, 
+but for a rate limiter registered using the `register` method you also need to apply it to a [specific route](#rate-limiting-scope).
+
+
+### Configure rate limiting {id="configure-rate-limiting"}
+
+In this section, we'll see how to configure rate limiting:
+
+1. (Optional) The `register` method allows you to specify a rate limiter name that can be used to
+   apply rate limiting rules to [specific routes](#rate-limiting-scope):
+   ```kotlin
+       install(RateLimit) {
+           register(RateLimitName("protected")) {
+               // ...
+           }
+       }
+   ```
+
+2. The `rateLimiter` method creates a rate limiter with two parameters: 
+   `limit` defines the bucket capacity, while `refillPeriod` specifies a refill period for this bucket.
+   A rate limiter in the example below allows handling 30 requests per minute:
+   ```kotlin
+   ```
+   {src="snippets/rate-limit/src/main/kotlin/com/example/Application.kt" include-lines="21-22,32"}
+
+3. (Optional) `requestKey` allows you to specify a function that returns a key for a request.
+   Requests with different keys have independent rate limits.
+   In the example below, the `login` [query parameter](requests.md#query_parameters) is a key 
+   used to distinguish different users:
```

Comment:
Please specify that keys should have good `equals` and `hashCode` implementations.